### PR TITLE
avr: drop GNU toolchain dependency

### PR DESCRIPTION
**Pull request by aykevl**
- Use compiler-rt and picolibc instead of avr-libc.
  - Use ld.lld instead of avr-ld (or avr-gcc).

This makes it much easier to get started with TinyGo on AVR because installing these extra tools (gcc-avr, avr-libc) can be a hassle. It also opens the door for future improvements such as ThinLTO.

There is a code size increase but I think it's worth it in the long run. The code size increase can hopefully be reduced with improvements to the LLVM AVR backend and to compiler-rt.

---

~Draft, because these changes depend on picolibc fixes for AVR.~ apparently updating picolibc was enough to get it to compile (see https://github.com/picolibc/picolibc/pull/276).

Needed for https://github.com/tinygo-org/tinygo/issues/2870.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 FROM golang:1.19 AS tinygo-llvm
 
 RUN apt-get update && \
-    apt-get install -y apt-utils make cmake clang-11 binutils-avr gcc-avr avr-libc ninja-build
+    apt-get install -y apt-utils make cmake clang-11 ninja-build
```
Comment (aykevl) [inline Dockerfile:5]: Unrelated to this PR, but why is `clang-11` needed?

**Comment by aykevl** [comment]
I should also note that with this PR, you'll need to run `make gen-device-avr` otherwise the compiled firmware won't work.

**Comment by deadprogram** [comment]
@aykevl this looks like might be a real error?

https://github.com/tinygo-org/tinygo/actions/runs/3394816244/jobs/5643870561#step:19:475

```
wrong cast for pointers in different address spaces(must be an address space cast)!
```

**Comment by aykevl** [comment]
Yes, that looks like a Clang bug. I've fixed a few such bugs in the past but apparently there are more.

**Review by dgryski — APPROVED** [review]
LGTM

**Comment by aykevl** [comment]
Note to self, this is the crash:

```
wrong cast for pointers in different address spaces(must be an address space cast)!
UNREACHABLE executed at /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:2041!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: llvm-build.dbg/bin/clang -Werror -Wall -std=gnu11 -D_COMPILING_NEWLIB -D_HAVE_ALIAS_ATTRIBUTE -DTINY_STDIO -DPOSIX_IO -D_IEEE_LIBM -D__OBSOLETE_MATH_FLOAT=1 -D__OBSOLETE_MATH_DOUBLE=0 -D_WANT_IO_C99_FORMATS -nostdlibinc -isystem /home/ayke/src/tinygo/tinygo/lib/picolibc/newlib/libc/include -I/home/ayke/src/tinygo/tinygo/lib/picolibc/newlib/libc/tinystdio -I/home/ayke/src/tinygo/tinygo/lib/picolibc/newlib/libm/common -I/home/ayke/.cache/tinygo/picolibc-avr-atmega1284p/include -c -Oz -gdwarf-4 -ffunction-sections -fdata-sections -Wno-macro-redefined --target=avr -fdebug-prefix-map=/tmp/tinygo4117165684/build-lib-picolibc=/tmp/tinygo-picolibc -mmcu=atmega1284p -mdouble=64 -o freopen.o /home/ayke/src/tinygo/tinygo/lib/picolibc/newlib/libc/tinystdio/freopen.c
1.      <eof> parser at end of file
2.      /home/ayke/src/tinygo/tinygo/lib/picolibc/newlib/libc/tinystdio/freopen.c:42:1: LLVM IR generation of declaration 'freopen'
3.      /home/ayke/src/tinygo/tinygo/lib/picolibc/newlib/libc/tinystdio/freopen.c:42:1: Generating code for declaration 'freopen'
 #0 0x0000aaaab985ed90 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/ayke/src/tinygo/tinygo/llvm-project/llvm/lib/Support/Unix/Signals.inc:573:3
 #1 0x0000aaaab985d150 llvm::sys::RunSignalHandlers() /home/ayke/src/tinygo/tinygo/llvm-project/llvm/lib/Support/Signals.cpp:104:18
 #2 0x0000aaaab97c2750 HandleCrash /home/ayke/src/tinygo/tinygo/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:76:5
 #3 0x0000aaaab97c2750 CrashRecoverySignalHandler(int) /home/ayke/src/tinygo/tinygo/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:389:62
 #4 0x0000ffff8e06e7cc (linux-vdso.so.1+0x7cc)
 #5 0x0000ffff8dae2790 (/usr/lib/libc.so.6+0x82790)
 #6 0x0000ffff8da9b6fc gsignal (/usr/lib/libc.so.6+0x3b6fc)
 #7 0x0000ffff8da878b0 abort (/usr/lib/libc.so.6+0x278b0)
 #8 0x0000aaaab97c970c (llvm-build.dbg/bin/clang+0x279970c)
 #9 0x0000aaaab9f0c9ac clang::QualType::operator->() const /home/ayke/src/tinygo/tinygo/llvm-project/clang/include/clang/AST/Type.h:789:22
#10 0x0000aaaab9f0c9ac (anonymous namespace)::ScalarExprEmitter::VisitCastExpr(clang::CastExpr*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:2046:48
#11 0x0000aaaab9f0b30c (anonymous namespace)::ScalarExprEmitter::Visit(clang::Expr*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:408:57
#12 0x0000aaaab9f0b518 (anonymous namespace)::ScalarExprEmitter::VisitCastExpr(clang::CastExpr*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:2036:23
#13 0x0000aaaab9f16f98 Visit /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:408:57
#14 0x0000aaaab9f16f98 (anonymous namespace)::ScalarExprEmitter::VisitBinAssign(clang::BinaryOperator const*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:4284:16
#15 0x0000aaaab9f0d0d8 Visit /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:408:57
#16 0x0000aaaab9f0d0d8 clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExprScalar.cpp:4874:13
#17 0x0000aaaab9eb29a4 llvm::PointerIntPair<llvm::Value*, 1u, bool, llvm::PointerLikeTypeTraits<llvm::Value*>, llvm::PointerIntPairInfo<llvm::Value*, 1u, llvm::PointerLikeTypeTraits<llvm::Value*> > >::PointerIntPair() /home/ayke/src/tinygo/tinygo/llvm-project/llvm/include/llvm/ADT/PointerIntPair.h:52:13
#18 0x0000aaaab9eb29a4 clang::CodeGen::RValue::RValue() /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGValue.h:39:7
#19 0x0000aaaab9eb29a4 clang::CodeGen::RValue::get(llvm::Value*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGValue.h:90:12
#20 0x0000aaaab9eb29a4 clang::CodeGen::CodeGenFunction::EmitAnyExpr(clang::Expr const*, clang::CodeGen::AggValueSlot, bool) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExpr.cpp:217:23
#21 0x0000aaaab9ec3678 clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGExpr.cpp:192:62
#22 0x0000aaaab9beff38 llvm::IRBuilderBase::GetInsertBlock() const /home/ayke/src/tinygo/tinygo/llvm-project/llvm/include/llvm/IR/IRBuilder.h:173:47
#23 0x0000aaaab9beff38 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGStmt.cpp:123:56
#24 0x0000aaaab9bf6924 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CGStmt.cpp:496:31
#25 0x0000aaaab9c46fe8 clang::CodeGen::CodeGenFunction::checkIfFunctionMustProgress() /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenFunction.h:568:9
#26 0x0000aaaab9c46fe8 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenFunction.cpp:1231:34
#27 0x0000aaaab9c56188 clang::CodeGen::CodeGenModule::getLangOpts() const /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenModule.h:720:51
#28 0x0000aaaab9c56188 clang::CodeGen::CodeGenFunction::getLangOpts() const /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenFunction.h:1996:66
#29 0x0000aaaab9c56188 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenFunction.cpp:1443:18
#30 0x0000aaaab9ca0a4c clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:5238:3
#31 0x0000aaaab9c9c5a4 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3515:47
#32 0x0000aaaab9ca3db0 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (.part.0) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:6097:37
#33 0x0000aaaaba7e9694 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/ModuleBuilder.cpp:188:67
#34 0x0000aaaaba7ded1c clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:234:7
#35 0x0000aaaababa9900 clang::ParseAST(clang::Sema&, bool, bool) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Parse/ParseAST.cpp:166:17
#36 0x0000aaaaba1c9a94 clang::FrontendAction::Execute() /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1037:21
#37 0x0000aaaaba176f0c llvm::Error::getPtr() const /home/ayke/src/tinygo/tinygo/llvm-project/llvm/include/llvm/Support/Error.h:271:51
#38 0x0000aaaaba176f0c llvm::Error::operator bool() /home/ayke/src/tinygo/tinygo/llvm-project/llvm/include/llvm/Support/Error.h:234:22
#39 0x0000aaaaba176f0c clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1035:42
#40 0x0000aaaaba275578 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:266:38
#41 0x0000aaaab7f1a92c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ayke/src/tinygo/tinygo/llvm-project/clang/tools/driver/cc1_main.cpp:250:40
#42 0x0000aaaab7f15590 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) /home/ayke/src/tinygo/tinygo/llvm-project/clang/tools/driver/driver.cpp:317:20
#43 0x0000aaaaba034cbc operator() /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Driver/Job.cpp:407:32
#44 0x0000aaaaba034cbc void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::'lambda'()>(long) /home/ayke/src/tinygo/tinygo/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:52
#45 0x0000aaaab97c28d0 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) /home/ayke/src/tinygo/tinygo/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:426:10
#46 0x0000aaaaba0356f4 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (.part.0) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Driver/Job.cpp:407:3
#47 0x0000aaaaba008798 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Driver/Compilation.cpp:200:22
#48 0x0000aaaaba008f14 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&, bool) const /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Driver/Compilation.cpp:254:5
#49 0x0000aaaaba014620 llvm::SmallVectorBase<unsigned int>::empty() const /home/ayke/src/tinygo/tinygo/llvm-project/llvm/include/llvm/ADT/SmallVector.h:73:47
#50 0x0000aaaaba014620 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) /home/ayke/src/tinygo/tinygo/llvm-project/clang/lib/Driver/Driver.cpp:1739:28
#51 0x0000aaaab7f18ecc clang_main(int, char**) /home/ayke/src/tinygo/tinygo/llvm-project/clang/tools/driver/driver.cpp:513:39
#52 0x0000ffff8da87b80 (/usr/lib/libc.so.6+0x27b80)
#53 0x0000ffff8da87c60 __libc_start_main (/usr/lib/libc.so.6+0x27c60)
#54 0x0000aaaab7f145f0 _start /build/glibc/src/glibc/csu/../sysdeps/aarch64/start.S:84:0
clang-15: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 15.0.0 (https://github.com/espressif/llvm-project 28b26ff95250371c59cf7a231acab8ef3191dc8b)
Target: avr
Thread model: posix
InstalledDir: /home/ayke/src/tinygo/tinygo/llvm-build.dbg/bin
clang-15: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-15: note: diagnostic msg: /tmp/freopen-437a65.c
clang-15: note: diagnostic msg: /tmp/freopen-437a65.sh
clang-15: note: diagnostic msg: 

********************
```

**Comment by aykevl** [comment]
Fix (hopefully): https://github.com/picolibc/picolibc/pull/369

**Comment by aykevl** [comment]
This seems to be fixed with https://github.com/picolibc/picolibc/pull/370. For now I've worked around it by commenting out the `freopen` source file. I doubt `freopen` works as-is anyway and I doubt anybody will want to use it, so I don't see any harm in doing that.

Next time we update picolibc, we can hopefully remove this little hack.

File: Dockerfile (hunk 1 above)
Comment (deadprogram) [reply Dockerfile:5]: That is a good question. I wonder if it actually is needed now?

**Comment by deadprogram** [comment]
I am particularly excited about this PR! Alongside #3264 it simplifies the installation considerably for people who want to put code on what are such common MCUs in the wild.

We should also be able to remove the avr dependencies from https://github.com/tinygo-org/tinyhci/blob/main/tools/docker/Dockerfile#L4 and https://github.com/tinygo-org/docker/blob/master/Dockerfile#L26 once this is merged.

**Pull request merged**